tx-compiler/src/compiler.rs

```diff
@@ -0,0 +1,299 @@
+        target_account_proc: Vec<NoteTarget>,
+    ) -> Result<NoteScript, TransactionError> {
+        let (note_script, code_block) = NoteScript::new(
+            note_script_ast,
+            &mut self.assembler,
+            &mut AssemblyContext::new(AssemblyContextType::Program),
+        )
+        .map_err(|_| TransactionError::CompileNoteScriptFailed)?;
+
+        let branches = collect_call_branches(&code_block);
+        for note_target in target_account_proc.into_iter() {
+            verify_note_account_compatibility(
+                note_script.hash(),
+                &branches,
+                &self.get_target_interface(note_target)?,
+            )?;
+        }
+
+        Ok(note_script)
+    }
+
+    /// Returns the account interface asscoiated with the provided [NoteTarget].
+    ///
+    /// # Errors
+    /// - If the account interface associated with the [AccountId] provided as a target can not be
+    ///   found in the `account_procedures` map.
+    fn get_target_interface(&self, target: NoteTarget) -> Result<Vec<Digest>, TransactionError> {
+        match target {
+            NoteTarget::AccountId(id) => self
+                .account_procedures
+                .get(&id)
+                .cloned()
+                .ok_or(TransactionError::AccountInterfaceNotFound(id)),
+            NoteTarget::Procedures(procs) => Ok(procs),
+        }
+    }
+
+    /// Compiles a transaction which executes the provided notes against the specified account.
+    ///
+    /// The account is assumed to have been previously loaded into this compiler.
+    pub fn compile_transaction(
+        &mut self,
+        account_id: AccountId,
+        notes: Vec<Note>,
+        tx_script: Option<ProgramAst>,
+    ) -> Result<CompiledTransaction, TransactionError> {
+        // Check if the account has been loaded into the [TransactionCompiler]
+        if !self.account_procedures.contains_key(&account_id) {
+            return Err(TransactionError::AccountInterfaceNotFound(account_id));
+        }
+
+        // Transaction must contain at least one input note
+        if notes.is_empty() {
+            return Err(TransactionError::NoNotesProvided);
+        }
+
+        // Create the [AssemblyContext] for compilation of the transaction program
+        let mut assembly_context = AssemblyContext::new(AssemblyContextType::Program);
+
+        // Create note tree
+        let note_program_tree =
+            self.create_note_program_tree(&account_id, &notes, &mut assembly_context)?;
+
+        // Create the transaction script [CodeBlock].  If no tx_script is provided we use a
+        // Noop span block.
+        let tx_script = match tx_script {
+            Some(tx_script) => self
+                .assembler
+                .compile_in_context(tx_script, &mut assembly_context)
+                .map_err(TransactionError::CompileTxScriptFailed)?,
+            None => CodeBlock::new_span(vec![Operation::Noop]),
+        };
```

Comment:
We should also probably try to verify whether `tx_script` can be executed against this account.